src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/FileBinding.java

```diff
@@ -48,53 +46,38 @@
         return FileCredentials.class;
     }
 
-    @Override public SingleEnvironment bindSingle(Run<?,?> build, FilePath workspace, Launcher launcher, TaskListener listener) throws IOException, InterruptedException {
-        FileCredentials credentials = getCredentials(build);
-        FilePath secrets = secretsDir(workspace);
-        String dirName = UUID.randomUUID().toString();
-        final FilePath dir = secrets.child(dirName);
-        dir.mkdirs();
-        secrets.chmod(/*0700*/448);
+    @Override protected final FilePath write(FileCredentials credentials, FilePath dir) throws IOException, InterruptedException {
         FilePath secret = dir.child(credentials.getFileName());
-        copy(secret, credentials);
-        if (secret.isDirectory()) { /* ZipFileBinding */
-            // needs to be writable so we can delete its contents
-            // needs to be executable so we can list the contents
-            secret.chmod(0700);
-        }
-        else {
-            secret.chmod(0400);
-        }
-        return new SingleEnvironment(secret.getRemote(), new UnbinderImpl(dirName));
+        secret.copyFrom(credentials.getContent());
+        secret.chmod(0400);
+        return secret;
     }
-    
-    private static class UnbinderImpl implements Unbinder {
 
+    @SuppressWarnings("unused")
+    @Deprecated
+    private static class UnbinderImpl implements Unbinder {
         private static final long serialVersionUID = 1;
-
         private final String dirName;
-        
-        UnbinderImpl(String dirName) {
+        private transient AbstractOnDiskBinding.UnbinderImpl delegate;
+
+        private UnbinderImpl(String dirName) {
             this.dirName = dirName;
         }
-        
-        @Override public void unbind(Run<?, ?> build, FilePath workspace, Launcher launcher, TaskListener listener) throws IOException, InterruptedException {
-            secretsDir(workspace).child(dirName).deleteRecursive();
-        }
-        
-    }
 
-    private static FilePath secretsDir(FilePath workspace) {
-        return tempDir(workspace).child("secretFiles");
-    }
-
-    // TODO 1.652 use WorkspaceList.tempDir
-    private static FilePath tempDir(FilePath ws) {
-        return ws.sibling(ws.getName() + System.getProperty(WorkspaceList.class.getName(), "@") + "tmp");
-    }
+        protected Object readResolve() {
```

Comment:
Indeed there are tests with paused `program.dat` but it gets difficult when a `node` is involved.